### PR TITLE
JAMES-3596  Collocate all server applications in server/apps folder

### PR DESCRIPTION
**Pull request by chibenwa**
TODO :

 - [x] cassandra-app +cassandra-app-ldap-tests
 - [x] distributed-app + distributed-app-ldap-tests & fix broken links
 - [x] jpa-smtp-app
 - [x] spring-app & fix broken links
 - [x] cli
 - [x] cli integration tests
 - [x] webadmin-cli
 - [x] jpa-app (requires guice module extraction)
 - [x] memory-app (requires guice module extraction)

### CODE REVIEW

**Comment by chibenwa** [comment]
The result is easier to review on the main branch than as a PR...

https://github.com/chibenwa/james-project/tree/JAMES-3596/server/apps

**Comment by chibenwa** [comment]
Cc @Arsnael if you could have a look at it...

Cc @jeantil I don't know if you will get the time (wish you the best!) but as you were involved on the ML discussions...

If people do not mind I would like to merge this timely as the scene is now well setted up for some magestic git conflicts drama...

**Comment by jeantil** [comment]
I did a superficial review since the PR is huge and PRs consisting mainly of file moves are hard to read :) 
( I am currenty in the eye of the cyclone the package has arrived but won't reach home before saturday)
The new organization looks fine to me. 
The only thing that got me surprised was the choice to split 
- cassandra-app-ldap-tests
- cli-integration-tests
- distributed-ldap-tests
Away from their respective modules (if I understand this change correctly). It puts testing modules in the apps/assembly directory which I find counter intuitive even if they are clearly marked as being tests.

Can you explain the rationale behind the split ? I'm not opposed but I would like to understand the why as it breaks expectations, and maybe ask that the why be explained in a readme in the apps dir so that people proposing new assemblies or manipulating assemblies would know if they need to provide a test module too. 
Appart from that I'm all for it once the CI clears

**Comment by chibenwa** [comment]
> The only thing that got me surprised was the choice to split
> cassandra-app-ldap-tests
> cli-integration-tests
> distributed-ldap-tests
> Away from their respective modules (if I understand this change correctly). It puts testing modules in the apps/assembly directory which I find counter intuitive even if they are clearly marked as being tests

Regarding cli-integration-tests there should be little problems merging it with CLI

Regarding ldap-tests we need to figure out the most efficient way to merge them in back to their original module. Since we use module choosing it should be doable.

This sounds like nice enhancements for follow up PRs...

> Can you explain the rationale behind the split ?

Bad reasons. For cli-integration tests, likely not mixing integration tests and unit tests... For LDAP this used to be a separate APP, then we kept separate modules because of configuration files on the class path... Problem that we likely have tools to solve now!

File: server/apps/cassandra-app-ldap-tests/pom.xml [hunk 1]
```diff
@@ -23,21 +23,33 @@
...
         <groupId>org.apache.james</groupId>
-        <artifactId>james-server-guice</artifactId>
+        <artifactId>james-server</artifactId>
         <version>3.7.0-SNAPSHOT</version>
-        <relativePath>../pom.xml</relativePath>
+        <relativePath>../../pom.xml</relativePath>
     </parent>
 
-    <artifactId>james-server-cassandra-ldap-guice</artifactId>
+    <artifactId>james-server-cassandra-app-ldap-tessts</artifactId>
```
Comment (Arsnael) [inline server/apps/cassandra-app-ldap-tests/pom.xml]:
```suggestion
    <artifactId>james-server-cassandra-app-ldap-tests</artifactId>
```

File: core/src/test/java/org/apache/james/core/UsernameTest.java [hunk 2]
```diff
@@ -148,9 +149,9 @@ void fromUsernameShouldThrowWhenTooLong() {
     }
 
     @Test
-    void fromUsernameShouldThrowWhenMultipleDomainDelimiter() {
-        assertThatThrownBy(() -> Username.of("aa@aa@aa"))
-            .isInstanceOf(IllegalArgumentException.class);
+    void arbitraryCharInLocalPart() {
+        assertThatCode(() -> Username.of("Whatever_I- want3$$$###_! as lons I do not give a domain"))
+            .doesNotThrowAnyException();
```
Comment (Arsnael) [inline core/src/test/java/org/apache/james/core/UsernameTest.java:154]: Why do you change a test for Username in this commit/PR? Mix up with something else maybe?

File: server/container/guice/mailbox-jpa/pom.xml [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    </parent>
+
+    <artifactId>james-server-guice-mailbox-jpa</artifactId>
+    <packaging>jar</packaging>
+
+    <name>Apache James :: Server :: JPA - guice injection</name>
+    <description>An advanced email server - JPA backend with guice injection</description>
+
+    <properties>
+        <jpa.includes>empty</jpa.includes>
```
Comment (Arsnael) [inline server/container/guice/mailbox-jpa/pom.xml:38]: What is this for?

File: server/apps/cassandra-app-ldap-tests/pom.xml (hunk 1 above)
Comment (Arsnael) [reply server/apps/cassandra-app-ldap-tests/pom.xml]: Solved in a later commit, don't bother!

File: core/src/test/java/org/apache/james/core/UsernameTest.java (hunk 2 above)
Comment (chibenwa) [reply core/src/test/java/org/apache/james/core/UsernameTest.java:154]: Sorry mistake I likely wrote some tests to answer a ML thread while running long compilations

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
(Rebased before merge, I just squashed the fixups)

**Comment by Arsnael** [comment]
```
10:51:21,756 [ERROR]   The project org.apache.james:james-server-guice-jmx:3.7.0-SNAPSHOT (/home/jenkins/workspace/james_ApacheJames_PR-487/server/container/guice/jmx/pom.xml) has 1 error

10:51:21,757 [ERROR]     'dependencies.dependency.version' for org.apache.james:james-server-mailetcontainer-camel:jar is missing. @ line 67, column 21

10:51:21,758 [ERROR]   

10:51:21,758 [ERROR]   The project org.apache.james:james-server-guice-mailet:3.7.0-SNAPSHOT (/home/jenkins/workspace/james_ApacheJames_PR-487/server/container/guice/mailet/pom.xml) has 2 errors

10:51:21,759 [ERROR]     'dependencies.dependency.version' for org.apache.james:james-server-mailetcontainer-camel:jar is missing. @ line 70, column 21

10:51:21,760 [ERROR]     'dependencies.dependency.version' for org.apache.camel:camel-support:jar is missing. @ line 87, column 21
```

**Comment by chibenwa** [comment]
https://github.com/apache/james-project/pull/492 removes the LDAP modules.

**Pull request merged**